File: libs/sdk-go/pkg/daytona/otel.go
// Copyright 2025 Daytona Platforms Inc.
// SPDX-License-Identifier: Apache-2.0

package daytona

import (
	"context"
	"fmt"
	"net/http"
	"sync"
	"time"
	"unicode"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetrichttp"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
	"go.opentelemetry.io/otel/metric"
	"go.opentelemetry.io/otel/propagation"
	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.26.0"
	"go.opentelemetry.io/otel/trace"
)

// otelState holds all OpenTelemetry infrastructure for the SDK.
type otelState struct {
	tracerProvider *sdktrace.TracerProvider
	meterProvider  *sdkmetric.MeterProvider
	tracer         trace.Tracer
	meter          metric.Meter

	mu         sync.Mutex
	histograms map[string]metric.Float64Histogram
}

// getHistogram returns a cached histogram for the given metric name, creating one if needed.
func (s *otelState) getHistogram(name string) (metric.Float64Histogram, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	if h, ok := s.histograms[name]; ok {
		return h, nil
	}

	h, err := s.meter.Float64Histogram(name,
		metric.WithUnit("ms"),
		metric.WithDescription(fmt.Sprintf("Duration of %s in milliseconds", name)),
	)
	if err != nil {
		return nil, err
	}
	s.histograms[name] = h
	return h, nil
}

// initOtel creates and configures the OpenTelemetry providers.
func initOtel(ctx context.Context) (*otelState, error) {
	res, err := resource.New(ctx,
		resource.WithAttributes(
			semconv.ServiceName("daytona-go-sdk"),
			semconv.ServiceVersion(Version),
		),
	)
	if err != nil {
		return nil, fmt.Errorf("failed to create otel resource: %w", err)
	}

	// Trace exporter (HTTP, gzip)
	traceExporter, err := otlptracehttp.New(ctx,
		otlptracehttp.WithCompression(otlptracehttp.GzipCompression),
	)
	if err != nil {
		return nil, fmt.Errorf("failed to create trace exporter: %w", err)
	}

	tp := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(traceExporter),
		sdktrace.WithResource(res),
	)

	// Metric exporter
	metricExporter, err := otlpmetrichttp.New(ctx,
		otlpmetrichttp.WithCompression(otlpmetrichttp.GzipCompression),
	)
	if err != nil {
		// Clean up trace provider on failure
		_ = tp.Shutdown(ctx)
		return nil, fmt.Errorf("failed to create metric exporter: %w", err)
	}

	mp := sdkmetric.NewMeterProvider(
		sdkmetric.WithReader(sdkmetric.NewPeriodicReader(metricExporter)),
		sdkmetric.WithResource(res),
	)

	// Set global propagator for W3C TraceContext
	otel.SetTextMapPropagator(propagation.TraceContext{})

	tracer := tp.Tracer("daytona-go-sdk")
	meter := mp.Meter("daytona-go-sdk")

	return &otelState{
		tracerProvider: tp,
		meterProvider:  mp,
		tracer:         tracer,
		meter:          meter,
		histograms:     make(map[string]metric.Float64Histogram),
	}, nil
}

// shutdownOtel flushes and shuts down both providers.
func shutdownOtel(ctx context.Context, state *otelState) error {
	if state == nil {
		return nil
	}

	var firstErr error
	if err := state.tracerProvider.Shutdown(ctx); err != nil {
		firstErr = err
	}
	if err := state.meterProvider.Shutdown(ctx); err != nil && firstErr == nil {
		firstErr = err
	}
	return firstErr
}

// otelTransport wraps an http.RoundTripper to inject trace context headers.
type otelTransport struct {
	base http.RoundTripper
}

func (t *otelTransport) RoundTrip(req *http.Request) (*http.Response, error) {
	// Inject traceparent / tracestate headers
	otel.GetTextMapPropagator().Inject(req.Context(), propagation.HeaderCarrier(req.Header))
	return t.base.RoundTrip(req)
}

// withInstrumentation wraps a function with OpenTelemetry span creation and duration metrics.
// When state is nil (OTel disabled), fn is called directly with zero overhead.
func withInstrumentation[T any](ctx context.Context, state *otelState, component, method string, fn func(ctx context.Context) (T, error)) (T, error) {
	if state == nil {
		return fn(ctx)
	}

	spanName := component + "." + method
	ctx, span := state.tracer.Start(ctx, spanName,
		trace.WithAttributes(
			attribute.String("component", component),
			attribute.String("method", method),
		),
	)
	defer span.End()

	start := time.Now()
	result, err := fn(ctx)
	duration := float64(time.Since(start).Milliseconds())

	status := "success"
	if err != nil {
		status = "error"
		span.RecordError(err)
	}

	metricName := toSnakeCase(spanName) + "_duration"
	if h, hErr := state.getHistogram(metricName); hErr == nil {
		h.Record(ctx, duration,
			metric.WithAttributes(
				attribute.String("component", component),
				attribute.String("method", method),
				attribute.String("status", status),
			),
		)
	}

	return result, err
}

// withInstrumentationVoid wraps an error-only function with OpenTelemetry instrumentation.
func withInstrumentationVoid(ctx context.Context, state *otelState, component, method string, fn func(ctx context.Context) error) error {
	_, err := withInstrumentation(ctx, state, component, method, func(ctx context.Context) (struct{}, error) {
		return struct{}{}, fn(ctx)
	})
	return err
}

// toSnakeCase converts a PascalCase or camelCase string to snake_case.
// Dots are replaced with underscores for Prometheus-friendly metric names.
func toSnakeCase(s string) string {
	result := make([]byte, 0, len(s)*2)
	for i, r := range s {
		if r == '.' {
			result = append(result, '_')
			continue
		}
		if unicode.IsUpper(r) && i > 0 {
			prev := rune(s[i-1])
			if prev != '.' && !unicode.IsUpper(prev) {
				result = append(result, '_')
			}
		}
		result = append(result, byte(unicode.ToLower(r)))
	}
	return string(result)
}

File: libs/sdk-rust/src/otel.rs
// Copyright 2019-2025 Daytona Labs Inc.
// SPDX-License-Identifier: Apache-2.0

#![allow(dead_code)]
// SPDX-License-Identifier: Apache-2.0

use crate::error::DaytonaError;
use opentelemetry::global;
use opentelemetry::trace::{Span, Status, Tracer};
use opentelemetry_otlp::SpanExporter;
use opentelemetry_sdk::trace::SdkTracerProvider;
use opentelemetry_sdk::Resource;

const SERVICE_NAME: &str = "daytona-sdk-rust";
const TRACER_NAME: &str = "daytona-sdk";

pub(crate) struct OtelState {
    tracer_provider: SdkTracerProvider,
}

impl OtelState {
    /// Initialize OTLP HTTP exporter, create tracer provider, and set as global.
    ///
    /// The exporter reads configuration from standard OpenTelemetry environment
    /// variables (`OTEL_EXPORTER_OTLP_ENDPOINT`, etc.) and defaults to
    /// `http://localhost:4318`.
    pub fn new() -> Result<Self, DaytonaError> {
        let exporter = SpanExporter::builder()
            .with_http()
            .build()
            .map_err(|e| DaytonaError::Network(format!("Failed to create OTLP exporter: {e}")))?;

        let tracer_provider = SdkTracerProvider::builder()
            .with_batch_exporter(exporter)
            .with_resource(Resource::builder().with_service_name(SERVICE_NAME).build())
            .build();

        global::set_tracer_provider(tracer_provider.clone());

        Ok(Self { tracer_provider })
    }

    /// Shutdown the tracer provider, flushing any pending spans.
    pub async fn shutdown(&self) -> Result<(), DaytonaError> {
        self.tracer_provider
            .shutdown()
            .map_err(|e| DaytonaError::Network(format!("Failed to shutdown tracer provider: {e}")))
    }
}

/// Instrument a function call with OpenTelemetry tracing.
///
/// If `otel` is `Some`, creates a span named `{component}.{method}` and
/// executes `f` within it. On error the span status is set accordingly.
/// If `otel` is `None`, `f` is called directly with no overhead.
pub(crate) async fn with_instrumentation<T, F, Fut>(
    otel: Option<&OtelState>,
    component: &str,
    method: &str,
    f: F,
) -> Result<T, DaytonaError>
where
    F: FnOnce() -> Fut,
    Fut: std::future::Future<Output = Result<T, DaytonaError>>,
{
    if otel.is_some() {
        let tracer = global::tracer(TRACER_NAME);
        let span_name = format!("{component}.{method}");
        let mut span = tracer.start(span_name);

        let result = f().await;

        if let Err(ref e) = result {
            span.set_status(Status::error(e.to_string()));
        }

        span.end();
        result
    } else {
        f().await
    }
}

/// Instrument a void function (no return value).
///
/// Convenience wrapper around [`with_instrumentation`] for `()` return type.
pub(crate) async fn with_instrumentation_void<F, Fut>(
    otel: Option<&OtelState>,
    component: &str,
    method: &str,
    f: F,
) -> Result<(), DaytonaError>
where
    F: FnOnce() -> Fut,
    Fut: std::future::Future<Output = Result<(), DaytonaError>>,
{
    with_instrumentation(otel, component, method, f).await
}

/// HTTP transport wrapper that injects `traceparent` header for
/// distributed trace context propagation.
pub(crate) struct OtelTransport<T> {
    inner: T,
}

#[allow(dead_code)]
impl<T> OtelTransport<T> {
    pub fn new(inner: T) -> Self {
        Self { inner }
    }

    pub fn inner(&self) -> &T {
        &self.inner
    }
}
